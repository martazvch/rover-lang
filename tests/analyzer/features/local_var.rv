var glob = 1
var period = "."
var loc2 = "sneaky"

fn main() {
    {
        var loc1 = 1.5
        var loc2 = 5
        var _loc3 = -.5

        print _loc3 + loc1

        {
            var loc1 = 6.
            var loc2 = 1.5

            print loc2 + loc1
            print loc1 + _loc3
        }

        print loc2
    }

    {
        var loc1 = "local"
        var loc2 = " var"

        print loc1 + loc2 + period

        loc2 = "assign local"
        {
            var loc2 = "sublocal"
            loc2 = "assign sub"
        }
        loc2 = loc1
    }
}

loc2 = "still glob"
