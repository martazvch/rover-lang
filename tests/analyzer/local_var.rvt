// Tests variable declaration type

code:
fn main() {
    {
        var loc1 = 1.5
        var loc2 = 5
        var _loc3 = -.5

        print _loc3 + loc1
    }
}

expect:
[Fn declaration, scope Global, index 0, return ImplicitVoid]
[Block pop count 3]
[Variable scope Local, index 1]
[Variable scope Local, index 2]
[Unary type float]
[Variable scope Local, index 3]

[Binop cast None, type float]
[Variable scope Local, index 3]
[Variable scope Local, index 1]

================

code:
fn main() {
    {
        var loc1 = 1.5
        var loc2 = 5
        var _loc3 = -.5

        print _loc3 + loc1

        {
            var loc1 = 6.
            var loc2 = 1.5

            print loc2 + loc1
            print loc1 + _loc3
        }

        print loc2
    }
}

expect:
[Fn declaration, scope Global, index 0, return ImplicitVoid]
[Block pop count 3]
[Variable scope Local, index 1]
[Variable scope Local, index 2]
[Unary type float]
[Variable scope Local, index 3]

[Binop cast None, type float]
[Variable scope Local, index 3]
[Variable scope Local, index 1]

[Block pop count 2]
[Variable scope Local, index 4]
[Variable scope Local, index 5]

[Binop cast None, type float]
[Variable scope Local, index 5]
[Variable scope Local, index 4]

[Binop cast None, type float]
[Variable scope Local, index 4]
[Variable scope Local, index 3]

[Variable scope Local, index 2]

================

code:
var glob = 1
var period = "."
var loc2 = "sneaky"

fn main() {
    {
        var loc1 = "local"
        var loc2 = " var"

        print loc1 + loc2 + period
    }
}

expect:
[Variable scope Global, index 0]
[Variable scope Global, index 1]
[Variable scope Global, index 2]
[Fn declaration, scope Global, index 3, return ImplicitVoid]

[Block pop count 2]
[Variable scope Local, index 1]
[Variable scope Local, index 2]

[Binop cast None, type str]
[Binop cast None, type str]
[Variable scope Local, index 1]
[Variable scope Local, index 2]
[Variable scope Global, index 1]

================

code:
var loc2 = "sneaky"

fn main() {
    {
        var loc1 = "local"
        var loc2 = " var"

        loc2 = "assign local"
        {
            var loc2 = "sublocal"
            loc2 = "assign sub"
        }
        loc2 = loc1
    }
}

loc2 = "still glob"

expect:
[Variable scope Global, index 0]
[Fn declaration, scope Global, index 1, return ImplicitVoid]

[Block pop count 2]
[Variable scope Local, index 1]
[Variable scope Local, index 2]
[Variable scope Local, index 2]

[Block pop count 1]
[Variable scope Local, index 3]
[Variable scope Local, index 3]

[Variable scope Local, index 1]
[Variable scope Local, index 2]

[Variable scope Global, index 0]

================

code:
var loc2 = "sneaky"
{
    var loc1 = "local"
    var loc1 = " var"
}

fn main() {}

error:
AlreadyDeclared, loc1
