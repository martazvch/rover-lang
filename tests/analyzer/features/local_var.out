[Variable declaration index: 0, scope: global]
    [Int 1]
[Variable declaration index: 1, scope: global]
    [String .]
[Variable declaration index: 2, scope: global]
    [String sneaky]
[Function declaration main]
    [Block pop count: 3, is_expr: false]
        [Variable declaration index: 0, scope: local]
            [Float 1.5]
        [Variable declaration index: 1, scope: local]
            [Int 5]
        [Variable declaration index: 2, scope: local]
            [Unary minus]
                [Float 0.5]
        [Print]
            [Binop type: add_float]
                [Variable index: 2, scope: local]
                [Variable index: 0, scope: local]
        [Block pop count: 2, is_expr: false]
            [Variable declaration index: 0, scope: local]
                [Float 6]
            [Variable declaration index: 1, scope: local]
                [Float 1.5]
            [Print]
                [Binop type: add_float]
                    [Variable index: 4, scope: local]
                    [Variable index: 3, scope: local]
            [Print]
                [Binop type: add_float]
                    [Variable index: 3, scope: local]
                    [Variable index: 2, scope: local]
        [Print]
            [Variable index: 1, scope: local]
    [Block pop count: 2, is_expr: false]
        [Variable declaration index: 0, scope: local]
            [String local]
        [Variable declaration index: 1, scope: local]
            [String  var]
        [Print]
            [Binop type: add_str]
                [Binop type: add_str]
                    [Variable index: 0, scope: local]
                    [Variable index: 1, scope: local]
                [Variable index: 1, scope: global]
        [String assign local]
        [Assignment index: 1, scope: local]
        [Block pop count: 1, is_expr: false]
            [Variable declaration index: 0, scope: local]
                [String sublocal]
            [String assign sub]
            [Assignment index: 2, scope: local]
        [Variable index: 0, scope: local]
        [Assignment index: 1, scope: local]
[String still glob]
[Assignment index: 2, scope: global]
